timemachine/md/barker.py

```diff
@@ -0,0 +1,61 @@
+import numpy as np
+
+
+class BarkerProposal:
+    def __init__(self, grad_log_q, proposal_sig=0.001):
+        """Robust gradient-informed proposal distribution.
+
+        Supports:
+        * sampling from proposal y ~ p(. | x)
+        * evaluation of proposal density p(y | x)
+
+        Compared to Langevin proposals, more robust to poor initialization and to poor choice of step size.
+
+        References
+        ----------
+        [Livingstone, Zanella, 2020]
+            The Barker proposal: combining robustness and efficiency in gradient-based MCMC
+            https://arxiv.org/abs/1908.11812
+        """
+        self.grad_log_q = grad_log_q
+        assert proposal_sig > 0
+        self.proposal_sig = proposal_sig
+
+    def _sample(self, x, gaussian_rvs, uniform_rvs):
+        """alg. 1"""
+        assert x.shape == gaussian_rvs.shape == uniform_rvs.shape
+
+        # sample from base kernel mu_sig
+        z = gaussian_rvs * self.proposal_sig
+
+        # p_xz = 1 / (1 + exp(-grad_x * z))
+        grad_x = self.grad_log_q(x)
+        log_p_xz = -np.logaddexp(0.0, -grad_x * z)
+        p_xz = np.exp(log_p_xz)
+
+        b_xz = np.sign(p_xz - uniform_rvs)
+
+        y = x + b_xz * z
+
+        return y
+
+    def sample(self, x):
+        """y ~ p(. | x)"""
+        gauss = np.random.randn(*x.shape)
+        unif = np.random.rand(*x.shape)
```

Comment:
Just to note another alternative, I personally like the explicit pattern that JAX uses, where all random functions must take in a random state as an additional argument. (Here it could look like adding an argument `rng: numpy.random.Generator`). I think it's a little more flexible than objects maintaining random state (e.g. simplifies running independent chains or parallelizing sampling) and more explicit (downside is there's a little more boilerplate required to pass the rng around).

Reply:
>Here it could look like adding an argument rng: numpy.random.Generator to the sample method)

That would also match https://github.com/proteneer/timemachine/blob/80b748034c63918e4b28027848d10ac1a1ded8dd/timemachine/integrator.py#L145-L158

>I think it's a little more flexible than objects maintaining random state (e.g. simplifies running independent chains or parallelizing sampling) and more explicit 

Good points

Reply:
Thanks for both suggestions.

In the interest of keeping the signature simple, applied @badisa 's suggestion of having the object maintain random state https://github.com/proteneer/timemachine/pull/978/commits/b8dc43b21bc110f8e080167a88f988e0659b0ca5

@mcwitt : more control is still available by providing own source of gaussian and uniform r.v.'s to `_sample`

Reply:
Thinking about this a bit more, having random states scattered around in different objects vs. a single global random state seems potentially a bit dangerous; for example, a user might reasonably expect that setting `np.random.seed(123)` would make the following code deterministic. But with hidden random state, we have to be sure to track down all objects that maintain their own random state and be sure to reset/reconstruct them too. Don't mean to derail this PR, but it seems to me worth returning to this in the future.

Reply:
>for example, a user might reasonably expect that setting np.random.seed(123) would make the following code deterministic.

Agreed that would be an easy mistake to make. But it would also be relatively easy to detect. (User could call the "deterministic" code twice and usually observe that the outputs aren't the same.)

A possible benefit of having the object maintain random state is that certain other kinds of subtler mistakes become harder to make. (E.g. it's ~impossible to accidentally reuse the same random seed on each iteration of a simulation loop `for _ in range(T): x = prop.sample(x)`.)

>Don't mean to derail this PR, but it seems to me worth returning to this in the future.

This sounds like a good area to consider more carefully in the future. Maybe we can take a more unified approach to random state in timemachine's python layer.

(I don't think we currently have a unified approach here: some objects maintain their own random state, some have random functions that accept numpy `rng`'s or jax rng `key`'s, and some sites reference global numpy random state.)

(And a user who is being extra careful about random state can still supply draws to the deterministic function `_sample(x, gaussian_rvs, uniform_rvs)`.)

Reply:
> E.g. it's ~impossible to accidentally reuse the same random seed on each iteration of a simulation loop

That's a good point, this does seem like a benefit of the current approach.

> I don't think we currently have a unified approach here

Agreed that we currently have mixed conventions. (Just wanted to express some reservations over "objects that use randomness maintain their own implicit random state" becoming the convention without further discussion.)

Sounds good to me to continue this discussion in another thread to not distract from this PR.

Reply:
>(Just wanted to express some reservations over "objects that use randomness maintain their own implicit random state" becoming the convention without further discussion.)

Agreed -- not obvious to me what's the best approach overall.

>Sounds good to me to continue this discussion in another thread to not distract from this PR.

Agreed! Migrated to https://github.com/proteneer/timemachine/issues/980